rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/rest/TbHttpClient.java

```diff
@@ -176,48 +151,53 @@ private boolean useAuth(String proxyUser, String proxyPassword) {
     }
 
-    void destroy() {
-        if (this.eventLoopGroup != null) {
-            this.eventLoopGroup.shutdownGracefully(0, 5, TimeUnit.SECONDS);
-        }
-    }
-
     public void processMessage(TbContext ctx, TbMsg msg) {
-        String endpointUrl = TbNodeUtils.processPattern(config.getRestEndpointUrlPattern(), msg);
-        HttpHeaders headers = prepareHeaders(msg);
-        HttpMethod method = HttpMethod.valueOf(config.getRequestMethod());
-        HttpEntity<String> entity;
-        if(HttpMethod.GET.equals(method) || HttpMethod.HEAD.equals(method) ||
-            HttpMethod.OPTIONS.equals(method) || HttpMethod.TRACE.equals(method) ||
-            config.isIgnoreRequestBody()) {
-            entity = new HttpEntity<>(headers);
-        } else {
-            entity = new HttpEntity<>(getData(msg), headers);
-        }
-
-        URI uri = buildEncodedUri(endpointUrl);
-        ListenableFuture<ResponseEntity<String>> future = httpClient.exchange(
-                uri, method, entity, String.class);
-        future.addCallback(new ListenableFutureCallback<ResponseEntity<String>>() {
-            @Override
-            public void onFailure(Throwable throwable) {
-                TbMsg next = processException(ctx, msg, throwable);
-                ctx.tellFailure(next, throwable);
+        try {
+            if (semaphore != null && !semaphore.tryAcquire(config.getReadTimeoutMs(), TimeUnit.MILLISECONDS)) {
+                ctx.tellFailure(msg, new RuntimeException("Timeout during waiting for reply!"));
+                return;
             }
 
-            @Override
-            public void onSuccess(ResponseEntity<String> responseEntity) {
-                if (responseEntity.getStatusCode().is2xxSuccessful()) {
-                    TbMsg next = processResponse(ctx, msg, responseEntity);
-                    ctx.tellSuccess(next);
-                } else {
-                    TbMsg next = processFailureResponse(ctx, msg, responseEntity);
-                    ctx.tellNext(next, TbRelationTypes.FAILURE);
-                }
+            String endpointUrl = TbNodeUtils.processPattern(config.getRestEndpointUrlPattern(), msg);
+            HttpMethod method = HttpMethod.valueOf(config.getRequestMethod());
+            URI uri = buildEncodedUri(endpointUrl);
+
+            RequestBodySpec request = webClient
+                    .method(method)
+                    .uri(uri)
+                    .headers(headers -> prepareHeaders(headers, msg));
+
+            if (HttpMethod.POST.equals(method) || HttpMethod.PUT.equals(method) ||
+                    HttpMethod.PATCH.equals(method) || HttpMethod.DELETE.equals(method) ||
+                    config.isIgnoreRequestBody()) {
+                request.body(BodyInserters.fromValue(getData(msg)));
             }
-        });
-        if (pendingFutures != null) {
-            processParallelRequests(future);
+
+            request
```

Comment:
Investigate is this executed in the same thread or in a thread pool? What thread pool is used?